### PR TITLE
IB/UD: fix ep_destroy

### PR DESCRIPTION
**Pull request by alex-mikheev**
@yosefe 
Calling uct_ep_destroy() will scedule a flush of outstanding data.
ep itself will stay and will be able to receive data until interface is destroyed.

In the future we may implement some kind of disconnect protocol
for UD endpoint.

To save memory dest_qpn field moved from base ud ep to verbs ud ep.

### CODE REVIEW

**Comment by shamisp** [comment]
autobot build

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1261/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/675/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
```
cc1plus: warnings being treated as errors
/autofs/nccs-svm1_proj/csc040/software/rhel5-x86_64/ucx-build/gtest-1.7.0/include/gtest/gtest.h: In function 'testing::AssertionResult testing::internal::CmpHelperEQ(const char*, const char*, const T1&, const T2&) [with T1 = int, T2 = unsigned int]':
/autofs/nccs-svm1_proj/csc040/software/rhel5-x86_64/ucx-build/gtest-1.7.0/include/gtest/gtest.h:1525:   instantiated from 'static testing::AssertionResult testing::internal::EqHelper<true>::Compare(const char*, const char*, const T1&, const T2&, typename testing::internal::EnableIf<(! testing::internal::is_pointer<T2>::value)>::type*) [with T1 = int, T2 = uint32_t]'
uct/test_ud.cc:380:   instantiated from here
/autofs/nccs-svm1_proj/csc040/software/rhel5-x86_64/ucx-build/gtest-1.7.0/include/gtest/gtest.h:1448: warning: comparison between signed and unsigned integer expressions
make[2]: *** [uct/gtest-test_ud.o] Error 1
make[2]: *** Waiting for unfinished jobs....
```

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1262/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1263/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/676/ for details (Mellanox internal link).

File: src/uct/ib/ud/base/ud_ep.c [hunk 1]
```diff
@@ -782,4 +785,24 @@ void uct_ud_ep_pending_purge(uct_ep_h ep_h, uct_pending_callback_t cb)
...
+     * new data
+     * In the future consider doin full fledged disconnect
+     * protocol. Kind of TCP (FIN/ACK). Doing this will save memory
+     * on the other hand active ep will need more memory to keep its state
+     * and such protocol will add extra complexity
+     */
+
+    ucs_trace_func("");
+    /* cancel user pending */
+    uct_ud_ep_pending_purge(ep, 0);
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.c]: pls change 0 to NULL

File: src/uct/ib/ud/base/ud_iface.h [hunk 2]
```diff
@@ -288,5 +288,17 @@ uct_ud_iface_progress_pending_tx(uct_ud_iface_t *iface)
     }
 }
 
+/* Go over all active eps and remove them. Do it this way because class destructors are not
+ * virtual 
+ */
+#define UCT_UD_IFACE_DELETE_EPS(iface, ep_type_t) \
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_iface.h]: pls use indentation, and start macro arguments with _

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1282/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/695/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
:+1: 
failure is unrelated (fedora23 compilation issue)

**Pull request merged**